src/ir/analysis/derive_default.rs

```diff
@@ -0,0 +1,301 @@
+            TypeKind::ObjCId |
+            TypeKind::ObjCSel |
+            TypeKind::ObjCInterface(..) |
+            TypeKind::Enum(..) => {
+                trace!("    types that always cannot derive Default");
+                self.insert(id)
+            }
+
+            TypeKind::Array(t, len) => {
+                if self.cannot_derive_default.contains(&t) {
+                    trace!("    arrays of T for which we cannot derive Default \
+                            also cannot derive Default");
+                    return self.insert(id);
+                }
+
+                if len <= RUST_DERIVE_IN_ARRAY_LIMIT {
+                    trace!("    array is small enough to derive Default");
+                    ConstrainResult::Same
+                } else {
+                    trace!("    array is too large to derive Default");
+                    self.insert(id)
+                }
+            }
+
+            TypeKind::ResolvedTypeRef(t) |
+            TypeKind::TemplateAlias(t, _) |
+            TypeKind::Alias(t) => {
+                if self.cannot_derive_default.contains(&t) {
+                    trace!("    aliases and type refs to T which cannot derive \
+                            Default also cannot derive Default");
+                    self.insert(id)
+                } else {
+                    trace!("    aliases and type refs to T which can derive \
+                            Default can also derive Default");
+                    ConstrainResult::Same
+                }
+            }
+
+            TypeKind::Comp(ref info) => {
+                assert!(
+                    !info.has_non_type_template_params(),
+                    "The early ty.is_opaque check should have handled this case"
+                );
+
+                if info.kind() == CompKind::Union {
+                    if self.ctx.options().unstable_rust {
+                        trace!("    cannot derive Default for Rust unions");
+                        return self.insert(id);
+                    }
+
+                    if ty.layout(self.ctx)
+                        .map_or(true,
+                                |l| l.opaque().can_trivially_derive_default(self.ctx, ())) {
+                        trace!("    union layout can trivially derive Default");
+                        return ConstrainResult::Same;
+                    } else {
+                        trace!("    union layout cannot derive Default");
+                        return self.insert(id);
+                    }
+                }
+
+                if self.ctx.lookup_item_id_has_vtable(&item.id()) {
+                    trace!("    comp with vtable cannot derive Default");
+                    return self.insert(id);
+                }
+
+                if info.needs_explicit_vtable(self.ctx, item) {
+                    trace!("    comp that needs vtable cannot derive Default");
+                    return self.insert(id);
+                }
```

Comment:
Is it ever the case that a type can need an explicit vtable but not be in the `has_vtable` set??

I think this case is redundant...